### PR TITLE
Improve speed of /admin/assets

### PR DESCRIPTION
**Pull request by Martchus**
Very WIP, idea:

* Serve data as JSON
* Cache data on cleanup and serve statically when
  displaying /admin/assets

---

See https://progress.opensuse.org/issues/35389

### CODE REVIEW

**Comment by coolo** [comment]
@richiejp possible to move the other/*ltp*txt assets into test results? That would remove a lot of pressure from this page, even with rendering static caches it's still 10 seconds :(

**Comment by Martchus** [comment]
So this should basically work now, including to retrigger the `limit_assets` Minion task. The cache file is created under `$OPENQA_BASEDIR/openqa/cache` which is created if it doesn't exist yet. There's only one version of the file kept at the same time.

The only outstanding problem which come to my mind right now is preforking. Likely I have to do some locking to prevent race conditions.

File: lib/OpenQA/Schema/ResultSet/Assets.pm [hunk 1]
```diff
@@ -273,6 +282,25 @@ END_SQL
         $group_info{$largest_group}->{picked} += $size;
     }
 
+    # produce cache file for /admin/assets
+    my $cache_file_path = status_cache_file;
+    try {
+        my $cache_file = Mojo::File->new($cache_file_path);
+        # ensure parent directory exists
+        $cache_file->dirname->make_path();
```
Comment (coolo) [inline lib/OpenQA/Schema/ResultSet/Assets.pm]: you can't create that file as geekotest in production - you need to create this in the package as writeable

Comment (Martchus) [reply lib/OpenQA/Schema/ResultSet/Assets.pm]: Seems like the directory is already packages - but it belongs to the worker package. So I better choose another path, maybe `$OPENQA_BASEDIR/openqa/webui/cache`?

**Comment by richiejp** [comment]
Huh? Can't this be solved with pagination?

But to answer your question; I don't think so because they are used in child tasks.

**Comment by Martchus** [comment]
I fixed the race-condition by limiting the GRU task to one at a time. If the assets page is accessed while cleanup it running it refuses to load showing an error message and a retry button.

> Huh? Can't this be solved with pagination?

We already have a pagination. You likely mean only serving the first page in the first place and do additional AJAX requests for further pages. We do that already for some tables (using server-side DataTables).

The problem here is that the information in the table is actually the outcome of the cleanup algorithm which is just some Perl data structure. So selecting ranges isn't as easy/fast as for a simple database query.

Of course we could move everything into the database so selecting ranges would be fast. But even then each query would take a little bit considering the huge amount of data and that all sorting and filtering is now done on the server-side. So when using the table one would have a small loading time after each action. This is likely not better than a longer loading time in the first place but almost no loading time after each action.

Besides, we also have the assets listed by group on the bottom of the page. Fitting that into a pagination would a lot of work.

So I still think serving a static/cached file is the best approach here. It can be implemented with reasonable effort and the loading time is ok.

> But to answer your question; I don't think so because they are used in child tasks.

Maybe multiple of those assets could be at least bundled together. I mean it is likely possible to put the ~ 80 `*.qcow2.txt` files into a single archive and only add that as an asset.

**Comment by coolo** [comment]
Hmm, now that you mention it - this doesn't sound too bad. Instead of filling a JSON, (re)fill a DB table that is only there to render /admin/assets - you can paginate the hell out of it.

**Comment by Martchus** [comment]
Especially doing the 'assets by group' at the bottom this way would be quite some work. So I consider this idea an further enhancement which can be built on top of what this PR already achieves. For instance using JSON in the DataTable and for the  'assets by group' would be required for server-side rendering as well.

**Comment by richiejp** [comment]
Maybe mark the asset as cleaned up with a timestamp, then finding what assets were cleaned up becomes a repeatable query and you are not juggling data between requests. Then have another GC job which eventually deletes the rows after some reasonably long period of time and just stores some scalar stats for debugging.

Postrgre SQL will easily handle it so long as you at least index the cleanup timestamp field.

I general I think you should paginate everything, we don't want unbounded queries on a monster like OpenQA.

>  I mean it is likely possible to put the ~ 80 *.qcow2.txt files into a single archive and only add that as an asset.

This is only a temporary solution, although we can do it without too much trouble.

**Comment by coolo** [comment]
The problem is not the list of assets - the problem is that the assignment to job group change every couple of hours.

File: openQA.spec [hunk 2]
```diff
@@ -247,6 +247,7 @@ install -m 0644 %{_sourcedir}/openQA.changes %{buildroot}%{_datadir}/openqa/publ
 #
 mkdir %{buildroot}%{_localstatedir}/lib/openqa/pool/1
 mkdir %{buildroot}%{_localstatedir}/lib/openqa/cache
+mkdir %{buildroot}%{_localstatedir}/lib/openqa/webui/cache
```
Comment (coolo) [inline openQA.spec]: I don't think this works. mkdir will fail unless you also create (and package) %{_localstatedir}/lib/openqa/webui

Comment (Martchus) [reply openQA.spec]: Fixed. I can try to build and install it on e212 to be sure the packaging works.

Comment (coolo) [reply openQA.spec]: do that

**Comment by Martchus** [comment]
Seems to work on e212 now (even though now all assets are cleaned and hence the table empty).

**Comment by Martchus** [comment]
However, there's one problem: It now only shows the state *before* the last cleanup. So we would actually need to run the cleanup in dry mode once more to know how it looks after. (Like we currently run the cleanup in dry mode each time when serving the page.)

Note that this problem would be the same for the server-side approach described in my comment from yesterday. That is because it wouldn't be more 'live' - just the place where the data is stored and queried from is different.

**Comment by coolo** [comment]
Well, what we see in limit_assets minion task are a lot of UPDATE calls. If those don't happen, but pure selects and calculation, I don't think there is a problem in doing it twice in the minion task.

**Comment by coolo** [comment]
(of course with the server side DB we would obviously see update calls, that wasn't my point :)

**Pull request merged**